### DISCUSSION TITLE
Is it possible to run a hook in the background?

### DISCUSSION THREAD

**Discussion by Zegnat** [pocketbase/pocketbase#5259] [Q&A]
I have a sometimes long-running task using `$http.send()` that I would like to run through a JS hook when a record has been added to the database.

My first thought was to use the `onModelAfterCreate`-hook, as that would allow me to run something after having data added to the database. But it looks like `onModelAfterCreate` is “blocking” (so to speak) and will not actually allow the database to continue with other tasks until it is done.

It seems like the lifecycle seems to be something like `saveRecord` is started. Within `saveRecord` it runs `onModelBeforeCreate`, executes against db, and then runs `onModelAfterCreate`, before ever returning `saveRecord`. I expected it to return `saveRecord` before it ran `onModelAfterCreate` and add the after create hook to some sort of internal queue.

The lifecycle confused me even more in the following example:

```javascript
onModelAfterCreate(({ model, dao }) => {
    dao.runInTransaction((db) => {
        const taskRunId = $security.randomString(16);
        const record = db.findRecordById(model.tableName(), model.getId());
        console.log(`[${taskRunId}] Starting long running task on ${record.id}`);
        sleep(90000);
        console.log(`[${taskRunId}] Ending long running task on ${record.id}`);
        db.deleteRecord(record);
    });
}, "items");

onAfterBootstrap(({ app }) => {
    const collection = app.dao().findCollectionByNameOrId("items");
    app.dao().runInTransaction((db) => {
        const records = db.findRecordsByExpr(collection.id);
        for (let i = 0; i < 5; i++) {
            const record = new Record(collection, {});
            db.saveRecord(record);
        }
    });
});
```

What I expected to happen:
1. On application start, we create 5 records, and these should all get committed to the database at the same time because of the transaction.
2. As soon as the records are committed 5 after creates should be queued to be run in the JSVM.
3. Because of the JSVM pool, all those hooks should be able to run simultaneously.

What actually seems to happen:
1. On application start, it starts the loop of record creation.
2. For every time it gets to `db.saveRecord()` it starts running the after creation hook synchronously.
3. The hook starts to run the slow task.
4. When the hook is done, we return to the loop, and finally move on to the next record.
5. We do not make use of the JSVM pool, and whatever the hook does becomes part of the main transaction.

Is there something in PocketBase that would enable me to implement my expected use flow, where hooks will run asynchronously on the data in the database?

**Comment by ganigeorgiev** [comment]
You could check https://github.com/pocketbase/pocketbase/discussions/5235.

**Reply by ganigeorgiev** [reply]
Also, even if you execute your hook's code concurrently, keep in mind that SQLite allows only a single writer at a time and having slow/long-running transactions (especially mixed read&write) should be avoided as this will prevent other writes and transactions to proceed.

**Reply by Zegnat** [reply]
Are you saying that the `onModelAfterCreate` hooks cannot run concurrently (each in their own JSVM), or are they running concurrently in my example but are blocking eachother because of my transactions?

My actual use-case should be able to work around having to put transactions in. I will try without it and see if I see more things run simultaneously.

**Reply by ganigeorgiev** [reply]
All hook handlers run synchronously as part of the **original started gorouitine** that it is triggering them.

`onModelAfterCreate` is triggered every time a db model is created using the `app.Dao`. The creation could happen either as part of a http request (each request is executed in its own goroutine) or as in your case triggered by the `saveRecord` call in the `onAfterBootstrap` hook.

It is a little outdated, but you can find some general info how the JSVM pb_hooks works in https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210.

I don't understand what is your actual use case to provide a suggestion but having blocking/long-running tasks wrapped in `runInTransaction` block should be avoided. This in general apply for other dbs too but for SQLite is especially important since the transaction lock is for the entire db (even with the new transaction/batch web APIs that will be available with the [next major release](https://github.com/pocketbase/pocketbase/discussions/4355), the batch transactions are limited both in their total execution time and the number of create/update/delete operations).
